processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -251,6 +247,99 @@ public static boolean convertSegment(File toConvert, File converted, IndexSpec i
     public MMappedIndex mapDir(File inDir) throws IOException;
   }
 
+  public static void validateRowValues(
+      Rowboat rb1,
+      IndexableAdapter adapter1,
+      Rowboat rb2,
+      IndexableAdapter adapter2
+  )
+  {
+    final int[][] dims1 = rb1.getDims();
+    final int[][] dims2 = rb2.getDims();
+    if (dims1.length != dims2.length) {
+      throw new SegmentValidationException(
+          "Dim lengths not equal %s vs %s",
+          Arrays.deepToString(dims1),
+          Arrays.deepToString(dims2)
+      );
+    }
+    final Indexed<String> dim1Names = adapter1.getDimensionNames();
+    final Indexed<String> dim2Names = adapter2.getDimensionNames();
+    for (int i = 0; i < dims1.length; ++i) {
+      final int[] dim1Vals = dims1[i];
+      final int[] dim2Vals = dims2[i];
+      final String dim1Name = dim1Names.get(i);
+      final String dim2Name = dim2Names.get(i);
+      final Indexed<String> dim1ValNames = adapter1.getDimValueLookup(dim1Name);
+      final Indexed<String> dim2ValNames = adapter2.getDimValueLookup(dim2Name);
+
+      if (dim1Vals.length != dim2Vals.length) {
+        if (dim1Vals.length == 0 || dim2Vals.length == 0) {
+          // Might be OK if one of them has null
+          if (dim1Vals.length == 0) {
+            final String dimValName = dim2ValNames.get(dim2Vals[0]);
+            if (!Strings.isNullOrEmpty(dimValName)) {
+              throw new SegmentValidationException(
+                  "Dim [%s] value [%s] is not null. %s vs %s",
+                  dim2Name,
+                  dimValName,
+                  Arrays.deepToString(dims1),
+                  Arrays.deepToString(dims2)
+              );
+            }
+          } else {
+            // dim2Vals.length must be 0
+            final String dimValName = dim1ValNames.get(dim1Vals[0]);
+            if (!Strings.isNullOrEmpty(dimValName)) {
+              throw new SegmentValidationException(
+                  "Dim [%s] value [%s] is not null. %s vs %s",
+                  dim1Name,
+                  dimValName,
+                  Arrays.deepToString(dims1),
+                  Arrays.deepToString(dims2)
+              );
+            }
+          }
+        } else {
+          throw new SegmentValidationException(
+              "Dim [%s] value lengths not equal %s vs %s",
+              dim1Name,
+              Arrays.deepToString(dims1),
```

Comment:
not sure we need to print out the full array. Given that we are only comparing lengths, just the lengths should be enough here. This might get unwieldy for large multi-value fields.